content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,84 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+- **Random inputs should be _truly_ (pseudo-)random**, i.e. they should be _dynamically_ generated on every run with the help of suitable library functions such as `Math.random()` in JavaScript or `rand()` in C. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
+- In some languages such as C/C++, **make sure that the RNGs are properly seeded with the current time at the beginning** using `srand(time(NULL))` or equivalent; otherwise, the same pseudorandom sequence will be generated on every run and your "random" inputs will devolve into fixed ones.
+- **The order of random tests should be non-deterministic** especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria). Not only should inputs be randomized, but their order as well. It should not be the case that, for example, 10 valid, random inputs are tested in the first step followed by 10 invalid ones. The order of tested scenarios should be unpredictable too, and one possible technique is to generate and collect a set of random inputs for all required scenarios and shuffle them.
+- Under some rare circumstances, it's allowed to use so called **randomized tests** instead of fully random ones. For some types of problems (for example, kata related to chess, or problems with small set of possible inputs) it's too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too), and shuffle them before the tests are run. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.
+- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
+- **Difficulty should be consistent between test runs.** When randomized ranges are very large, it becomes possible for some users to receive many small, easy inputs while other users receive the exact opposite. Make sure that your random tests are built in a way which minimizes the chances for such a situation to occur. If you want to test difficult inputs, split the test cases into a set of easy ones and a set of difficult ones, and test them separately.
+- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
```

Comment:
```suggestion
- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
```